File: C/CWE-22.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void process_file(const char *filename) {
    char buffer[256];
    char filepath[512] = "/var/data/";  // Assume restricted directory

    // Concatenate filename to the base directory
    strcat(filepath, filename);

    // Attempt to open the file
    FILE *file = fopen(filepath, "r");
    if (file) {
        // Read data from the file
        while (fgets(buffer, sizeof(buffer), file)) {
            printf("%s", buffer);
        }
        fclose(file);
    } else {
        printf("File not found or cannot be opened.\n");
    }
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <filename>\n", argv[0]);
        return 1;
    }

    process_file(argv[1]);

    return 0;
}

File: C#/CWE-22.cs
using System;
using System.IO;

class Program
{
    static void Main(string[] args)
    {
        if (args.Length != 1)
        {
            Console.WriteLine("Usage: Program.exe <filename>");
            return;
        }

        string filename = args[0];
        string basePath = "/var/data/"; // Assume restricted directory

        // Construct full file path
        string filePath = basePath + filename;

        try
        {
            // Attempt to read the file
            string content = File.ReadAllText(filePath);
            Console.WriteLine(content);
        }
        catch (FileNotFoundException)
        {
            Console.WriteLine("File not found or cannot be opened.");
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error accessing file: " + ex.Message);
        }
    }
}
